File: decode/js/rosetta-image-decode/src/internal/jpeg.ts
import { DecodeError } from "../errors.js";
import type { DecodedImage } from "../types.js";
import { sniffJpegDimensions } from "./dimensionSniff.js";
import { checkDimensions } from "./limits.js";
import { loadWasmModule, resolvePackageWasmUrl } from "./loadWasm.js";

// @jsquash/jpeg: community-maintained mozjpeg WASM fork of squoosh's jpeg codec.
// Uses the same mozjpeg C library as Rust's mozjpeg-sys, ensuring byte-exact output.
import { init, default as jsquashDecode } from "@jsquash/jpeg/decode.js";

let wasmInitPromise: Promise<void> | null = null;

async function ensureWasmInit(): Promise<void> {
  if (wasmInitPromise) return wasmInitPromise;

  wasmInitPromise = (async () => {
    // Locate the WASM binary tolerantly: in Node, resolve via the package
    // graph (handles hoisted node_modules layouts); in browsers, fall back
    // to the bundler-relative URL pattern that esbuild/vite/webpack rewrite
    // to point at the emitted asset path.
    const wasmUrl = await resolvePackageWasmUrl(
      "@jsquash/jpeg/codec/dec/mozjpeg_dec.wasm",
      "../../node_modules/@jsquash/jpeg/codec/dec/mozjpeg_dec.wasm",
      import.meta.url,
    );
    const wasmModule = await loadWasmModule(wasmUrl);
    await init(wasmModule);
  })();

  // Reset on rejection so a transient failure (e.g., file missing during a
  // hot-reload) doesn't permanently poison every future call with the same
  // rejected promise.
  wasmInitPromise.catch(() => { wasmInitPromise = null; });

  return wasmInitPromise;
}

/**
 * Scan the JPEG stream to detect a 4-component color space (CMYK/YCCK)
 * before passing to the WASM decoder.  mozjpeg calls exit(1) on CMYK
 * which would terminate the Node.js process, so we must bail out early.
 *
 * Strategy: find the first SOF marker and read the number of components.
 * JPEG 4-component images are always CMYK or YCCK — never RGBA.
 */
function detectCmyk(bytes: Uint8Array): boolean {
  if (bytes.length < 4) return false;
  // Skip SOI (FF D8)
  let i = 2;
  while (i + 3 < bytes.length) {
    if (bytes[i] !== 0xff) break;
    const marker = bytes[i + 1]!;
    if (marker === 0xd9) break; // EOI

    // Stand-alone markers (no length field): RST0-RST7, SOI
    if (marker === 0xd8 || (marker >= 0xd0 && marker <= 0xd7)) {
      i += 2;
      continue;
    }

    if (i + 3 >= bytes.length) break;
    const length = ((bytes[i + 2]! << 8) | bytes[i + 3]!) >>> 0;
    if (length < 2) break;

    // SOF markers: C0 baseline, C1 extended, C2 progressive, C3 lossless, etc.
    if (
      marker === 0xc0 || marker === 0xc1 || marker === 0xc2 || marker === 0xc3 ||
      marker === 0xc5 || marker === 0xc6 || marker === 0xc7 ||
      marker === 0xc9 || marker === 0xca || marker === 0xcb ||
      marker === 0xcd || marker === 0xce || marker === 0xcf
    ) {
      // SOF layout: [length(2), precision(1), height(2), width(2), components(1), ...]
      // i points at 0xFF; segment data starts at i+2 (after marker bytes)
      // Offset from i: +4=precision, +5,+6=height, +7,+8=width, +9=nComponents
      if (i + 9 < bytes.length) {
        const numComponents = bytes[i + 9]!;
        if (numComponents === 4) return true;
      }
      break;
    }

    i += 2 + length;
  }
  return false;
}

export async function decodeJpeg(bytes: Uint8Array): Promise<DecodedImage> {
  // Pre-decode CMYK check — mozjpeg calls exit(1) on CMYK images which kills Node
  if (detectCmyk(bytes)) {
    throw new DecodeError("unsupportedFeature", "jpeg", "CMYK color space is not supported");
  }

  // Sniff dimensions from the first SOF marker BEFORE the WASM decoder
  // allocates the raster. Spec §3.1 requires this ordering: a file whose
  // header declares a huge canvas must error with imageTooLarge rather
  // than allocating then complaining.
  const sniffed = sniffJpegDimensions(bytes);
  if (sniffed) {
    checkDimensions(sniffed.width, sniffed.height, "jpeg");
  }

  await ensureWasmInit();

  // Node.js Buffer.buffer may be a shared pool with non-zero byteOffset;
  // always slice to get a fresh ArrayBuffer aligned to the payload.
  const arrayBuffer = bytes.buffer.slice(
    bytes.byteOffset,
    bytes.byteOffset + bytes.byteLength,
  ) as ArrayBuffer;

  let result: ImageData;
  try {
    result = await jsquashDecode(arrayBuffer);
  } catch (e: any) {
    const msg: string = e?.message ?? String(e);
    throw new DecodeError("corruptInput", "jpeg", `JPEG decode failed: ${msg}`);
  }

  if (!result) {
    throw new DecodeError("corruptInput", "jpeg", "JPEG decode returned null");
  }

  checkDimensions(result.width, result.height, "jpeg");

  // mozjpeg returns RGBA; strip the alpha channel to produce packed RGB
  const pixels = result.width * result.height;
  const rgb = new Uint8Array(pixels * 3);
  const rgba = result.data;
  for (let i = 0, di = 0, si = 0; i < pixels; i++, si += 4, di += 3) {
    rgb[di] = rgba[si]!;
    rgb[di + 1] = rgba[si + 1]!;
    rgb[di + 2] = rgba[si + 2]!;
  }

  return {
    width: result.width,
    height: result.height,
    data: rgb,
    channels: 3,
    format: "jpeg",
  };
}

File: decode/go/imagedecode/jpeg.go
package imagedecode

import (
	"github.com/wmetcalf/rosetta-image-decode/go/imagedecode/internal/libjpeg"
)

func decodeJpeg(b []byte) (DecodedImage, error) {
	// Check for CMYK/YCCK before invoking the decoder.
	if isCmyk(b) {
		return DecodedImage{}, newError(UnsupportedFeature, Jpeg, true, "CMYK color space")
	}

	// Use libjpeg.Header to get dimensions and guard against decompression bombs.
	width, height, err := libjpeg.Header(b)
	if err != nil {
		return DecodedImage{}, newError(CorruptInput, Jpeg, true, "libjpeg header: "+err.Error())
	}
	if err := checkDimensions(width, height, Jpeg); err != nil {
		return DecodedImage{}, err
	}

	// DecodeRGB uses TJFLAG_ACCURATEDCT (JDCT_ISLOW) to match PIL output byte-exact.
	// libjpeg-turbo forces YCbCr→RGB conversion and expands grayscale to 3-channel RGB.
	decoded, err := libjpeg.DecodeRGB(b)
	if err != nil {
		return DecodedImage{}, newError(CorruptInput, Jpeg, true, "libjpeg decode: "+err.Error())
	}
	return DecodedImage{
		Width:    decoded.Width,
		Height:   decoded.Height,
		Data:     decoded.Pixels,
		Channels: RGB,
		Format:   Jpeg,
	}, nil
}

// jpegDimensions scans JPEG markers to find the first SOF (Start Of Frame) marker
// and returns the image width and height. Returns (0, 0, false) if not found.
// SOF layout: FF Cn [length 2 bytes] [precision 1 byte] [height 2 bytes] [width 2 bytes] ...
//
// Currently unused; kept for parity with the other ports' JPEG helpers and
// available for future limit/dimension-sniff use.
//nolint:unused
func jpegDimensions(b []byte) (width, height int, ok bool) {
	for i := 0; i+1 < len(b); i++ {
		if b[i] != 0xFF {
			continue
		}
		mk := b[i+1]
		switch mk {
		case 0xC0, 0xC1, 0xC2, 0xC9, 0xCA:
			// SOF marker: precision(1) height(2) width(2) at i+4..i+8
			if i+8 < len(b) {
				h := int(b[i+4])<<8 | int(b[i+5])
				w := int(b[i+6])<<8 | int(b[i+7])
				return w, h, true
			}
		case 0x00, 0xFF:
			// stuffed byte or padding; continue
		case 0xD8, 0xD9:
			// SOI / EOI — no length field
		default:
			if i+3 < len(b) {
				length := int(b[i+2])<<8 | int(b[i+3])
				i += 1 + length
			}
		}
	}
	return 0, 0, false
}

// isCmyk scans JPEG markers to find the first SOF (Start Of Frame) marker
// and returns true if the image has 4 components (CMYK or YCCK).
func isCmyk(b []byte) bool {
	// JPEG markers are 0xFF followed by a marker byte.
	// SOF markers: C0 (baseline), C1 (extended sequential), C2 (progressive),
	//              C9 (arithmetic sequential), CA (arithmetic progressive).
	// SOF layout: FF Cn [length 2 bytes] [precision 1 byte] [height 2 bytes] [width 2 bytes] [num_components 1 byte]
	// num_components is at offset +9 from the start of the FF byte.
	for i := 0; i+1 < len(b); i++ {
		if b[i] != 0xFF {
			continue
		}
		mk := b[i+1]
		switch mk {
		case 0xC0, 0xC1, 0xC2, 0xC9, 0xCA:
			// SOF marker found; num_components is at i+9
			if i+9 < len(b) {
				return b[i+9] == 4
			}
		case 0x00, 0xFF:
			// 0x00 = stuffed byte (not a real marker), 0xFF = padding; skip one byte
			// (the outer loop increments i too, so we land on next byte correctly)
		case 0xD8, 0xD9:
			// SOI / EOI — no length field; continue scanning
		default:
			// All other markers have a 2-byte length field after the marker byte.
			// Skip past this marker: i+1 (marker byte) + 1 + length = i+1+1+length
			if i+3 < len(b) {
				length := int(b[i+2])<<8 | int(b[i+3])
				i += 1 + length // outer loop adds 1 more
			}
		}
	}
	return false
}
